pkg/reconciler/knativeeventing/knativeeventing.go

```diff
@@ -99,7 +97,7 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, ke *v1alpha1.KnativeEven
 		r.ensureFinalizerRemoval,
 		r.install,
 		r.checkDeployments,
-		r.deleteObsoleteResources,
+		r.deleteObsoleteResources(ctx, ke),
```

Comment:
What is the benefit of keeping deleteObsoleteResources as a stage?
If we only runs `installed.Filter(mf.None(mf.In(*manifest))).Delete()`, after the Stages, what is bad about that?

Reply:
You could, as long as you create the installed manifest before you execute the stages. I thought it was cleaner to put that logic in the `deleteObsoleteResources` function instead of `ReconcileKind`.

Reply:
What about this?
Change this line into `r.deleteObsoleteResources,`.
Change the function `deleteObsoleteResources` into 
```
func (r *Reconciler) deleteObsoleteResources(ctx context.Context, manifest *mf.Manifest, instance v1alpha1.KComponent) error {
   ...
   ...
   return installed.Filter(mf.None(mf.In(*manifest))).Delete()
}
```

Reply:
In your example, where is `installed` coming from. You can't create it "inside the pipeline" because one of the previous stages might set `instance.Status.Version` to something other than what it was when it was originally passed to `ReconcileKind`.

Reply:
installed, err := r.installed(ctx, instance)
I omit all the lines above.

Reply:
If you do that, you'll never delete any obsolete resources. The version of your installed manifest will be identical to your target manifest because `r.install` updates `status.version` upon successful completion: https://github.com/knative/operator/blob/9a134e09fe836907592f8d2cafb4f2b491eb493b/pkg/reconciler/common/install.go#L50

Because both the installed version *and* the target version are both determined by looking at the CR instance, you must calculate both using the _same_ instance _before_ any stages are executed that might mutate the instance.

You could argue that calculating the installed version in `ReconcileKind` is more clear and obviates the `NoOp` stages, but I liked having just the stages in there, and traded a bit of complexity for that.

Before we go that route, I'll see if I can add an e2e test to this PR that demonstrates the condition.

Reply:
Got it. Thx.